### PR TITLE
Fix incorrect propagation of OSErrors in create code

### PR DESCRIPTION
**Pull request by enkore**
Pulling a thumbdrive during a backup now produces the correct result:

```
Traceback (most recent call last):
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 81, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 247, in do_create
    create_inner(archive, cache)
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 219, in create_inner
    read_special=args.read_special, dry_run=dry_run)
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 275, in _process
    status = archive.process_file(path, st, cache, self.ignore_inode)
  File "/home/mabe/Projekte/_oss/borg/borg/archive.py", line 594, in process_file
    chunks.append(cache.add_chunk(self.key.id_hash(chunk), chunk, self.stats))
  File "/home/mabe/Projekte/_oss/borg/borg/cache.py", line 365, in add_chunk
    self.repository.put(id, data, wait=False)
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 467, in put
    segment, offset = self.io.write_put(id, data)
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 706, in write_put
    fd = self.get_write_fd(raise_full=raise_full)
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 600, in get_write_fd
    self._write_fd = open(self.segment_filename(self.segment), 'ab')
OSError: [Errno 5] Input/output error: '/run/media/mabe/00B524275BF67732/repo/data/0/9'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 1569, in main
    exit_code = archiver.run(args)
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 1506, in run
    return args.func(args)
  File "/home/mabe/Projekte/_oss/borg/borg/archiver.py", line 81, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 86, in __exit__
    self.close()
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 181, in close
    self.lock.release()
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 291, in release
    self._roster.modify(EXCLUSIVE, REMOVE)
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 203, in modify
    roster = self.load()
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 182, in load
    data = json.load(f)
  File "/home/mabe/.pyenv/versions/3.5.1/lib/python3.5/json/__init__.py", line 265, in load
    return loads(fp.read(),
OSError: [Errno 5] Input/output error

Platform: Linux tiger 4.5.7-1-ck #1 SMP PREEMPT Wed Jun 8 08:42:43 EDT 2016 x86_64 
Linux: arch  
Borg: 1.0.4.dev46+ngbc0a838.d20160627  Python: CPython 3.5.1
PID: 27050  CWD: /home/mabe/Projekte/_oss/borg/issue1138
sys.argv: ['/home/mabe/.pyenv/versions/3.5.1/bin/borg', 'create', '/run/media/mabe/00B524275BF67732/repo::testbreaking', '...............', '-svp']
SSH_ORIGINAL_COMMAND: None

zsh: exit 2     borg create /run/media/mabe/00B524275BF67732/repo::testbreaking  -svp
```

Tried it a few times, logic seems solid. I don't think I missed a spot, but that's what we do code reviews for, isn't it? ;)

_Rather curious_: ntfs-3g journaling doesn't seem to be very reliable. After a few pulls with no repository corruption I get an EIO on open(2) on a data file and a rather ugly ntfs-3g message:

> Jun 27 20:59:35 tiger ntfs-3g[27314]: Trying to read non-allocated mft records (83 > 82): Nicht erlaubter Seek (_seek not permissible_)
> Jun 27 20:59:35 tiger ntfs-3g[27314]: Trying to read non-allocated mft records (83 > 82): Nicht erlaubter Seek

Even root can't touch that inode anymore, not even rm the dirent. I'll see whether Windows chkdsk can fix that, or whether I have to reformat my thumb drive?!

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
hmm, how about stat()? And the bsdflags on linux also do OS calls.

**Comment by enkore** [comment]
The stat is done in Archiver._process and try/except'ed directly (l 259).

Linux BSD-Flags is only in master but we'll need to remember to fix that there as well, yes.

**Comment by ThomasWaldmann** [comment]
As a side note: in the changeset I am currently working on (in-file checkpoints), I unified the stdin and file reading code, so the duplication you have seen will be removed there.

**Comment by ThomasWaldmann** [comment]
Changeset looks good. As usual, the traceback isn't pretty / helpful for non-pythonista eyes.

As we can expect such exceptions and it is not our fault / nothing to debug for us, maybe you could subclass your exception class from Error (the one without traceback), so users will get a more pretty error msg.

**Comment by enkore** [comment]
Hmm, not sure about that. The exception causing the traceback is OSError unchanged, so we could catch and remove the traceback in main(), but since not every OSError carries a file name (most don't ... those concerning open(2) normally do) it can be hard to know *where* the error is in the system. There the traceback can help.

**Comment by ThomasWaldmann** [comment]
ah, right, confused that. %-)

File: borg/archive.py [hunk 1]
```diff
@@ -45,6 +46,37 @@
...
+def input_io_iter(iterator):
+    while True:
+        try:
+            with input_io():
+                item = next(iterator)
+        except StopIteration:
+            return
+        yield item
+
+
```
Comment (ThomasWaldmann) [inline borg/archive.py]: in case it is likely that we need this outside archive module later: move it to helpers.py?

File: borg/remote.py [hunk 2]
```diff
@@ -287,6 +290,8 @@ def fetch_from_cache(args):
                         type, msgid, error, res = unpacked
                         if msgid in self.ignore_responses:
                             self.ignore_responses.remove(msgid)
+                            if error:
+                                handle_error(error, res)
```
Comment (ThomasWaldmann) [inline borg/remote.py]: did you analayse the impact this missing code had? for changelog / commit comment.

Comment (enkore) [inline borg/remote.py]: The way I understand it had a similar impact as the original issue. An error (during e.g. a put()) would still be visible in the log, but not cause a program abort and may have stored/corrupted data.

Re. changelog I think it's the same bug in terms of description, that was caused two different ways for local and remote repos each.

**Comment by ThomasWaldmann** [comment]
merge?

**Pull request merged**